Setup Poetry, Tox, Travis, and add PB800 as an example

Eric-Arellano opened the pull request:
This provides the main setup for this new repository, as follows:

## Poetry

Currently, we use Flit for Pex and Pipenv for `rbe-token-server`. However, we decided to switch over Pex and `rbe-token-server` to use Poetry. Why? Flit can't be used for `rbe-token-server`, which is an app, and we want to be able to use the same tool for all of our repos.

Poetry also has substantial community momentum currently.

## Tox

We use Tox to simplify testing every supported Python version and to align with Pex.

## Travis

We stick with Travis because all other projects use Travis and we're familiar with it.

## Flake8 plugin setup

This follows the architecture of [`flake8-2020`](https://github.com/asottile/flake8-2020/blob/master/flake8_2020.py) in its definition of a `Visitor(ast.NodeVisitor)` and `Plugin` class. This plugin is developed by one of the maintainers of Flake8 and is only 4 months old, so is a good example of how to write Flake8 plugins.  

After reviewing several plugins, including `flake8-2020`, [`flake8-bugbears`](https://github.com/PyCQA/flake8-bugbear), [`flake8-eradicate`](https://github.com/sobolevn/flake8-eradicate), [`flake8-print`](https://github.com/JBKahn/flake8-print), and [`flake8-comprehensions`](https://github.com/adamchainz/flake8-comprehensions), it appears Flake8 plugins conventionally define all their logic in one module and one `ast.NodeVisitor` class, in contrast to `pantsbuild.contrib...checker` having a distinct file for each check. If this ends up not scaling well, we'll revisit alternative architectures, but for now, we follow the convention of other projects.

### Testing

We use `pytest-flake8dir`, which is a utility for testing Flake8 plugins. This makes it easy for us to write mock files to a tempdir and then run Flake8 against them and to capture the output.

Eric-Arellano committed:
Setup Poetry and PNT800

jsirois commented on README.md:
````diff
@@ -35,3 +35,37 @@ If using without Pants, run `flake8 your_module.py` [as usual](http://flake8.pyc
+
+```bash
+$ poetry build
+$ pex flake8 dist/flake8_pantsbuild-0.1.0-py2.py3-none-any.whl --entry-point flake8 --interpreter-constraint='CPython>=3.6' --output-file flake8.pex
````
`pex flake8 . --entry-point ...` should be enough, pex now knows PEP-517/PEP-518.

Eric-Arellano replied:
TIL. Thanks!

Eric-Arellano committed:
Fix interpreter constraints

Eric-Arellano committed:
New approach to testing: pytest-flake8dir

Also fix isort config and install Black

Eric-Arellano committed:
Add tox and Travis config

Eric-Arellano commented on poetry.lock:
```diff
@@ -0,0 +1,692 @@
+[[package]]
```
NB: This lock file is only used for developers of this library. It is in no way included in the distribution - the distribution will use the floating requirements of `pyproject.toml`.

The lockfile is nice for development becuase it means transitive dependencies won't change on us overnight like has happened with `more-itertools` breaking `pytest` in Pants.

jsirois replied:
If dependencies do change they should break though since they will in the wild. The convenience for developers is just a delay of the eventual break out in the real world.

Eric-Arellano commented on pyproject.toml:
```diff
@@ -0,0 +1,35 @@
+name = "flake8-pantsbuild"
+version = "0.1.0"
+description = "Various Flake8 lints used by the Pants project and its users."
+authors = ["Eric Arellano <ericarellano@me.com>"]
```
Wasn't sure what to put here. Should we include the `pantsbuild@gmail.com` email? Credit the original authors of the `checker` wheel upon which this is based?

I wanted to leave this field off, but Poetry requires it. There's also an optional `maintainers` field.

jsirois replied:
Perhaps follow pex's lead?

jsirois replied:
https://github.com/pantsbuild/pex/blob/master/pyproject.toml#L1-L8

Eric-Arellano replied:
Thanks for pointing to this! Updated.

Eric-Arellano commented on flake8_pantsbuild.py:
```diff
@@ -0,0 +1,51 @@
+    from importlib_metadata import version
+
+
+PNT800 = (
```
Note that the original code was `T800`. Flake8 strongly encourages plugin authors to use 3 letters for their codes due to reduce ambiguity and avoid possible name clashes with other plugins. 3 letter codes weren't possible until recently.

`PNT` seemed the best acronym for `Pants` that I could think of, but also considered `PAN`.

jsirois replied:
Since this plugin has ~nothing to do with Pants, how about deriving the code from pantsbuild instead, which suggests PB.

Eric-Arellano replied:
Ah, great suggestion!

Eric-Arellano commented on pyproject.toml:
```diff
@@ -0,0 +1,35 @@
+PNT = "flake8_pantsbuild:Plugin"
+
+[tool.poetry.dependencies]
+python = "~2.7 || ^3.5"
```
Poetry supports the conventional `<`, `<=`, `==` etc. style of version strings, along with the NPM style used here. The `~` means support any patch version with this minor version and `^` means support any minor version with this major version. See https://python-poetry.org/docs/dependency-specification/#version-constraints.

jsirois replied:
Feeling you have to explain this NPM style seems to tip the hat not to use the foreign style and instead stick to the conventions. Perhaps better would be to explain why you didn't want to stick to the conventions instead of explaining the syntax. My guess is you have a good reason.

Eric-Arellano replied:
Fair, went with conventional version strings because my reason isn't good enough to justify the possible confusion. Thanks!

jsirois commented on README.md:
````diff
@@ -35,3 +35,38 @@ If using without Pants, run `flake8 your_module.py` [as usual](http://flake8.pyc
+### To publish
+
+```bash
+$ poetry build
````
Presumably `poetry publish` is enough.

Eric-Arellano replied:
Negative, but `poetry --build publish` would work. https://python-poetry.org/docs/libraries/#publishing-to-pypi

jsirois commented:
> whereas with Flit we would need to declare the dependencies on flake8 and importlib_metadata in both pyproject.toml and tox.ini.

This is not true. If you used flit you'd just move dev-dependencies out of pyproject.toml into tox.ini. The production dependencies of the distribution would be built in by tox when it called out to the pyproject.toml isolated build.

Eric-Arellano committed:
Review feedback

- Use PB instead of PNT
- Change author to "Pantsbuild developers"
- Use conventional version strings

jsirois commented:
As taled about offline, I'm fine with moving all out pantsbuold repos to poetry, but there is alot of justification for this that is simply invalid. The valid argument seems to be rbe-token-server, which _does_ require dep management unlike this lib. That then pulls every other project in.

If you agree, perhaps trim down the PR description significantly, most of the justification is not valid fwict and is distracting.

Eric-Arellano commented:
> This is not true. If you used flit you'd just move dev-dependencies out of pyproject.toml into tox.ini. The production dependencies of the distribution would be built in by tox when it called out to the pyproject.toml isolated build.

Oh, TIL. Cool! I agree, then, that the main motivation would be `rbe-token-server` which is a bit of an outlier in it being an app whereas every other repository we have is a library.

I'll try out using this with Flit given that you don't need to duplicate the production dependencies.

Eric-Arellano committed:
Merge the changes from the flit experiment, but still use poetry

We are going to end up using Flit, only need to use Poetry until https://github.com/takluyver/flit/pull/305 is published.

Eric-Arellano commented:
This is now ready for re-review. I experimented with Flit and agree that it's the way to go. However, we can't do this yet because of https://github.com/takluyver/flit/pull/305, which is fixed but not released.

I'd like to land this change with Poetry for now, so that we can make progress on the rest of the plugin. We need this PR to be landed before adding any new lints.

Eric-Arellano commented:
Actually, there's an argument for sticking with Poetry after all. 

I agree with John that the original approach was wrong to explicitly use Poetry's dependency management here, e.g. checking in `poetry.lock` was a bad practice because this is a library and not an application. 

I didn't realize, though, that Poetry can work just like Flit. This current PR is essentially the same as if we were to use Flit, minus slightly different entries in `pyproject.toml`, slightly different CLI invocations to publish to PyPI, and declaring the version and name in `pyproject.toml` instead of src code. Currently, we ignore Poetry's dependency management and can continue to do so.

So, if Poetry and Flit would be almost the same, why bother with Poetry? The main reason is so that we can update `rbe-token-server` (an application) to use Poetry instead of Pipenv, whereas `rbe-token-server` cannot use Flit because Flit does not have dependency management for applications. It'd be great if we could use the same tool for this repo, Pex, _and_ `rbe-token-server`. Another reason is that Poetry has had substantial momentum in the Python community recently and many are proposing consolidating around Poetry.* For example, our CI works fine with Poetry, but when using Flit as the build tool, the Python 2 shard fails due to the module name issue _and_ the Py35-Py37 shards fail because it can't find a `setup.py` file, which seems to be a bug.

*When it's Pipenv vs. Poetry, there seems to be a consensus to use Poetry. There are very few comparisons between Poetry vs. Flit, however, so this isn't a particularly compelling reason.

--

@jsirois what do you think? If you agree to Poetry over Flit, I am happy to make the changes to swap over `rbe-token-server` and pex to Poetry.

jsirois commented:
> So, if Poetry and Flit would be almost the same, why bother with Poetry? The main reason is so that we can update rbe-token-server (an application) 

My apologies for not being clear. I thought we had discussed exactly this point offline which I found to be trumping justification for poetry; ie: since we need it for one project we should use it for all.

SGTM

jsirois commented on pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+[build-system]
+# TODO: switch to Flit once it releases a fix for https://github.com/takluyver/flit/pull/305. See
```
I think this comment can now be killed.

jsirois commented on tox.ini:
```diff
@@ -0,0 +1,46 @@
+[tox]
+isolated_build = true
+minversion: 3.14.1
```
Here and on line 44 `:` is used instead of `=` as the kv separator. It would be good to be uniform.

Eric-Arellano replied:
Good catch. 

(I really don't like ini..there's no standard for it and so many "gotchas". One of my side projects coming up is to add support for `pants.toml`.)

jsirois commented on tox.ini:
```diff
@@ -0,0 +1,46 @@
+[flake8]
+max_line_length: 100
+extend_ignore =
+    E203
```
Which line triggers this? Afaict all `:` in python code above not embedded in strings or comments are not preceded by whitespace.

Eric-Arellano replied:
Huh, not an issue anymore. Cool. I think it was necessary when using Flit and I carried it over unnecessarily.

Eric-Arellano committed:
Review feedback

Eric-Arellano commented:
Thanks for the review, John!
